nncf/experimental/tensorflow/nncf_network.py

```diff
@@ -0,0 +1,150 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from typing import Union, Dict, Tuple, List, Any
+import itertools
+
+import tensorflow as tf
+
+from nncf.tensorflow.layers.operation import NNCFOperation
+from nncf.experimental.tensorflow.context import get_current_context
+from nncf.experimental.tensorflow.graph.transformations.commands import TFTargetPoint
+from nncf.experimental.tensorflow.patch_tf import Hook
+
+
+InputSignature = Union[tf.TensorSpec, Dict[str, tf.TensorSpec], Tuple[tf.TensorSpec, ...], List[tf.TensorSpec]]
+
+
+def _add_names_to_input_signature(input_signature: InputSignature):
+    xs = tf.nest.flatten(input_signature)
+    ys = []
+    for i, spec in enumerate(xs):
+        ys.append(
+            tf.TensorSpec.from_spec(
+                spec,
+                name=spec.name if spec.name else f'input_{i}'
+            )
+        )
+
+    return tf.nest.pack_sequence_as(input_signature, ys)
+
+
+class NNCFNetwork(tf.keras.Model):
+    """
+    Wraps the Keras model.
+    """
+
+    def __init__(self,
+                 model: tf.keras.Model,
+                 input_signature: InputSignature,
+                 **kwargs):
+        """
+        Initializes the NNCF network.
+
+        :param model: Keras model.
+        :param input_signature: Input signature of the moodel.
+        """
+        super().__init__(**kwargs)
+        self._model = model
+        self._input_signature = _add_names_to_input_signature(input_signature)
+
+        self.__dict__['_pre_hooks'] = {}  # type: Dict[str, List[Hook]]
```

Comment:
please leave a comment why it's needed in that way (about overriden __setattr__ in tf.keras.layers.Layer, tracking object and not saving pre_hooks)
it would be helpful for developers

Reply:
Done